keps/sig-api-machinery/2855-automated-stored-versions/README.md

```diff
@@ -0,0 +1,581 @@
+
+- A migration for one of the CRDs owned by the cert provisioner is started.
+- A new version of the cert provisioner is deployed that is not compatible with the CRD being migrated and introduces a new stored version.
+- The cert provisioner upgrade fails because of the admission webhook and is no longer able to provision new certificates.
+- The certificate for ksvm expires.
+
+In the situation described above, the cluster administrator would need to manually revert the cert provision upgrade to proceed.
+
+#### Alternative Approaches
+
+- Instead of relying on an annotation to identify CRDs currently being migrated, the admission webhook could check for the existence of a non-completed `StorageVersionMigration` resource.
+
+### Risks and Mitigations
+
+- It is possible that the admission webhook introduced in this KEP may prevent changes to a CRD's `spec.versions[*].storage` fields if the migrating annotation is not successfully removed. Update operations against a CRD will also fail if the admission webhook is unavailable. Cluster Administrators can resolve this issue by deleting or modifying the validating admission webhook.
+
+## Design Details
+
+### Kube Storage Version Migrator Changes
+
+- When reconciling a `StorageVersionMigration` CR, the [kube-storage-version-migrator] must start by adding the `ksvm.sigs.k8s.io/migrating=true` annotation to the CRD being migrated.
+- When a migration is successful, the migrated CRD's `status.storedVersion` is patched to only include the new stored version.
+- After a successful or failed migration, the [kube-storage-version-migrator] removes the `ksvm.sigs.k8s.io/migrating=true` annotation from the migrated CRD. 
+- Should the `StorageVersionMigration` CR be deleted, the kube-storage-version-migrator] should remove the `ksvm.sigs.k8s.io/migrating=true` annotation from the CRD.
+
+### Proposed Admission Webhook
+
+As described earlier, the admission webhook would prevent update operations to any CRD's `spec.versions[*].storage` fields when the `ksvm.sigs.k8s.io/migrating=true` annotation is present. An example of the ValidationWebhookConfiguration can be found below:
+
+```yaml=
+apiVersion: admissionregistration.k8s.io/v1beta1
+  kind: ValidatingWebhookConfiguration
+  metadata:
+    name: crd.kube-storage-version-migrator.io
+  webhooks:
+  - name: crd.kube-storage-version-migrator.io
+    clientConfig:
+      service:
+        namespace: <deployed namespace>
+        name: webhooks
+       path: <TBD>
+      caBundle: <KUBE_CA_HERE>
+    rules:
+    - operations:
+      - Update
+      apiGroups:
+      - "apiextensions.k8s.io"
+      apiVersions:
+      - "v1" # Exclude v1beta1 as it is no longer served in K8s 1.22: https://kubernetes.io/docs/reference/using-api/deprecation-guide/#customresourcedefinition-v122
+      resources:
+      - customresourcedefinitions
+    failurePolicy: fail
+```
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
```

Comment:
> For this enhancement, I think you'll need to describe how you'll test this. 

Agreed.

>Since it sits out of tree, I'm not entirely clear on the test plan. e2e test on the kube-storage-version-migrator repo I suspect? Does it currently have the concept of alpha feature (not on by default) tests?

I believe that the tests should exist within the ksvm repo. The repo does not have a concept of alpha feature tests, and can introduce a flag that enables alpha tests.

I will also include some tests that are defined as acceptance criteria. Some tests that immediately jump to mind:
- The `ksvm.sigs.k8s.io/migrating=true` annotation is properly added to CRDs during migration.
- Changes to a CRD's `spec.versions[*].storage` fields are rejected during a migration.
- `ksvm.sigs.k8s.io/migrating=true` annotationis properly removed from CRDs post migration. 
- A CRD's stored version field is properly updated post migration.